### PR TITLE
Adapt `pallet-contracts` to WeightV2

### PR DESCRIPTION
**Pull request by athei**
A few changes were necessary to make `pallet-contracts` compatible with the new weight.

### Remove `ContractAccessWeight`

This was a workaround used to emulate the huge PoV impact generated by loading a contract. This is now replaced by setting the `proof_size` to the size of the loaded contract. This is an underestimation but the whole system is just in place to prevent the worst offenders. Charging for anything else is missing anyways and will come later.

### Return the whole `Weight` struct from dry runs

Dry runs now include the `proof_size` so that clients can use this to learn about how much `proof_size` a transaction will consume.

### Fix inter contract calls

`seal_call` and `seal_instantiate` allow only setting 1D weight. Eventually we need to add new versions of those. But the old versions need fixing, too. Current behavior is to set `proof_size` to `0` which is not a good default. It will prevent the callee to do anything useful. This PR sets the `proof_size` to "inherit" which allows it to use all of the remaining proof size from the overarching limit.

### Fix compatibility extrinsics

When adding the 2D weight we also added `*_old_weight` compat extrinsics to allow in-storage extrinsics to be still decoded. However, they also set `proof_size` to `0` which makes them useless. This PR sets some sensible default so they continue working. We also removed duplication in docs and code from them.

cumulus companion: https://github.com/paritytech/cumulus/pull/1726
Closes https://github.com/paritytech/substrate/issues/12410

### CODE REVIEW

**Comment by ascjones** [comment]
Closes https://github.com/paritytech/substrate/issues/12410

**Review by ascjones — APPROVED** [review]
LGTM, though I would really like to see some tests for the `proof_size` accounting

File: frame/contracts/src/tests.rs [hunk 1]
```diff
@@ -824,6 +823,8 @@ fn deploy_and_call_other_contract() {
 		// Drop previous events
 		initialize_block(2);
 
+		println!("--------------");
```
Comment (lexnv) [inline frame/contracts/src/tests.rs:826]: Nit: is this intended?

File: frame/contracts/src/lib.rs [hunk 2]
```diff
@@ -1287,4 +1162,12 @@ where
 	fn min_balance() -> BalanceOf<T> {
 		<T::Currency as Inspect<AccountIdOf<T>>>::minimum_balance()
 	}
+
+	/// Convert a 1D Weight to a 2D weight.
+	///
+	/// Used by backwards compatible extrinsics. We cannot just set the proof to zero
+	/// or an old `Call` will just fail.
+	fn compat_weight(gas_limit: OldWeight) -> Weight {
+		Weight::from(gas_limit).set_proof_size(u64::from(T::MaxCodeLen::get()) * 2)
```
Comment (lexnv) [inline frame/contracts/src/lib.rs:1171]: Dumb question: Is the `MacCodeLen * 2` always a sufficient approximation of gas?

**Review by lexnv — APPROVED** [review]
LGTM! Thanks for fixing this!

File: frame/contracts/src/lib.rs (hunk 2 above)
Comment (athei) [reply frame/contracts/src/lib.rs:1171]: Depends on how many contract you call and how big they are in a single transaction. This is just a ad-hoc solution to support in-storage `Call`s that might exist in a `Scheduler` of some runtime. This is very niche. You should use the dry-run to come up with a proper value and don't rely on this.

File: frame/contracts/src/tests.rs (hunk 1 above)
Comment (athei) [reply frame/contracts/src/tests.rs]: Nope. Some leftover. Thanks.

**Comment by athei** [comment]
bot merge

**Comment by athei** [comment]
bot merge

**Pull request merged**